sparse/coo/core.py

```diff
@@ -1572,28 +1560,303 @@ def _keepdims(original, new, axis):
+    overall mask, e.g. None and full slices.
+
+    Parameters
+    ----------
+    indices : tuple
+        The indices to the array.
+    shape : tuple[int]
+        The shape of the array.
+
+    Returns
+    -------
+    indices : tuple
+        The filtered indices.
+
+    Examples
+    --------
+    >>> _prune_indices((None, 5), (10,)) # None won't affect the mask
+    [5]
+    >>> _prune_indices((slice(0, 10, 1),), (10,)) # Full slices don't affect the mask
+    []
+    """
+    indices = [idx for idx in indices if idx is not None]
+    i = 0
+    for idx, l in zip(indices[::-1], shape[::-1]):
+        if not isinstance(idx, slice):
+            break
+
+        if idx.start == 0 and idx.stop == l and idx.step == 1:
+            i += 1
+            continue
+
+        if idx.start == l - 1 and idx.stop == -1 and idx.step == -1:
+            i += 1
+            continue
+
+        break
+    if i != 0:
+        indices = indices[:-i]
+    return indices
+
+
+@numba.jit(nopython=True)
+def _get_mask(coords, indices):  # pragma: no cover
+    """
+    Gets the mask for the coords given the indices in slice format.
+
+    Parameters
+    ----------
+    coords : np.ndarray
+        The coordinates of the array.
+    indices : np.ndarray
+        The indices in the form of slices.
+
+    Returns
+    -------
+    mask : np.ndarray
+        The starts and stops in the mask.
+    is_slice : bool
+        Whether or not the array represents a continuous slice.
+    """
+    starts = [0]
+    stops = [coords.shape[1]]
+    n_matches = coords.shape[1]
+
+    i = 0
+    while i < len(indices):
+        n_current_slices = _get_slice_len(indices[i]) * len(starts) + 2
+        if n_current_slices * np.log(n_current_slices / max(len(starts), 1)) > n_matches:
+            break
+
+        starts, stops, n_matches = _get_mask_pairs_inner(starts, stops, coords[i], indices[i])
+
+        i += 1
+
+    starts, stops = _join_adjacent_pairs(starts, stops)
+
+    if i == len(indices) and len(starts) == 1:
+        return np.array([starts[0], stops[0]]), True
+
+    mask = _cat_pairs(starts, stops)
+
+    while i < len(indices):
+        mask = _filter_mask(coords[i], mask, indices[i])
+        i += 1
+
+    return np.array(mask, dtype=np.intp), False
+
+
+@numba.jit(nopython=True)
+def _get_slice_len(idx):  # pragma: no cover
+    """
+    Get the number of elements in a slice.
+
+    Parameters
+    ----------
+    idx : np.ndarray
+        A (3,) shaped array containing start, stop, step
+
+    Returns
+    -------
+    n : int
+        The length of the slice.
+
+    Examples
+    --------
+    >>> idx = np.array([5, 15, 5])
+    >>> _get_slice_len(idx)
+    2
+    """
+    start, stop, step = idx[0], idx[1], idx[2]
+
+    if step > 0:
+        return (stop - start + step - 1) // step
+    else:
+        return (start - stop - step - 1) // (-step)
+
+
+@numba.jit(nopython=True)
+def _get_mask_pairs_inner(starts_old, stops_old, c, idx):  # pragma: no cover
```

Comment:
It looks like there are a number of functions like this that are only used once.  I'm somewhat inclined to just inline these directly into where they are called rather than construct a new function.  I personally find it difficult to follow a new codebase if I have to jump between several functions.  Giant long functions are also unpleasant of course, but I find them easier than those same lines of code split out to many places in the same file.

Reply:
This seems to happen a number of times here, and I suspect it is partially what is partially what is causing me to be unable to review this PR efficiently.  I wonder if the explanation could be handled as well with comments.

Reply:
I did it that way at first, but apparently Numba has some unfixed bugs with type inference if I inline these into their parent functions. If I keep the logic unchanged and inline these functions, Numba compilation fails as it fails to infer the type of some of the lists.

Reply:
If it helps, bottom-up reviewing may be better in this case than top-down.

Reply:
I've merged at least two functions. I tried to remove another one, but that unfortunately tanks the performance.